src/runtime/security/negotiation_service.cpp

```diff
@@ -44,10 +48,43 @@ void negotiation_service::on_negotiation_request(negotiation_rpc rpc)
         return;
     }
 
-    server_negotiation *srv_negotiation =
-        static_cast<server_negotiation *>(rpc.dsn_request()->io_session->get_negotiation());
+    server_negotiation *srv_negotiation = nullptr;
+    {
+        zauto_read_lock l(_lock);
+        srv_negotiation =
+            static_cast<server_negotiation *>(_negotiations[rpc.dsn_request()->io_session].get());
+    }
+
+    dassert(srv_negotiation != nullptr,
+            "negotiation is null for msg: {}",
+            rpc.dsn_request()->rpc_code().to_string());
     srv_negotiation->handle_request(rpc);
 }
 
+void negotiation_service::on_rpc_connected(rpc_session *session)
+{
+    std::unique_ptr<negotiation> nego = security::create_negotiation(session->is_client(), session);
+    nego->start();
+    {
+        zauto_write_lock l(_lock);
+        _negotiations[session] = std::move(nego);
```

Comment:
Is it safe to add an assert to check the session is not exist in the map?

Reply:
I think it's ok to not add assert here. If a rpc session connects twice, we should add a new negotiation, and the old one will be destroyed when `_negotiations[session] = std::move(nego)`